src/uucore/src/lib/features/checksum.rs

```diff
@@ -335,35 +299,184 @@ fn determine_regex(
+                },
+            )),
         }
+    } else {
+        Ok(caps.name("checksum").unwrap().as_str().to_string())
     }
+}
 
-    Err(
-        ChecksumError::NoProperlyFormattedChecksumLinesFound(get_filename_for_output(
-            filename,
-            input_is_stdin,
-        ))
-        .into(),
-    )
+/// Returns a reader that reads from the specified file, or from stdin if `filename_to_check` is "-".
+fn get_file_to_check(
+    filename_to_check: &str,
+    ignore_missing: bool,
+    res: &mut ChecksumResult,
+) -> Option<Box<dyn Read>> {
+    if filename_to_check == "-" {
+        Some(Box::new(stdin())) // Use stdin if "-" is specified in the checksum file
+    } else {
+        match File::open(filename_to_check) {
+            Ok(f) => {
+                if f.metadata().ok()?.is_dir() {
+                    show!(USimpleError::new(
+                        1,
+                        format!("{}: Is a directory", filename_to_check)
+                    ));
+                    None
+                } else {
+                    Some(Box::new(f))
+                }
+            }
+            Err(err) => {
+                if !ignore_missing {
+                    // yes, we have both stderr and stdout here
+                    show!(err.map_err_context(|| filename_to_check.to_string()));
+                    println!("{}: FAILED open or read", filename_to_check);
+                }
+                res.failed_open_file += 1;
+                // we could not open the file but we want to continue
+                None
+            }
+        }
+    }
+}
+
+/// Returns a reader to the list of checksums
+fn get_input_file(filename_input: &OsStr, input_is_stdin: bool) -> UResult<Box<dyn Read>> {
+    if input_is_stdin {
+        Ok(Box::new(stdin())) // Use stdin if "-" is specified
+    } else {
+        match File::open(filename_input) {
+            Ok(f) => {
+                if f.metadata()?.is_dir() {
+                    Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        format!("{}: Is a directory", filename_input.to_string_lossy()),
+                    )
+                    .into())
+                } else {
+                    Ok(Box::new(f))
+                }
+            }
+            Err(_) => Err(io::Error::new(
+                io::ErrorKind::Other,
+                format!(
+                    "{}: No such file or directory",
+                    filename_input.to_string_lossy()
+                ),
+            )
+            .into()),
+        }
+    }
+}
+
+/// Extracts the algorithm name and length from the regex captures if the algo-based format is matched.
+fn get_algo_name_and_length(
+    caps: &regex::Captures,
+    is_algo_based_format: bool,
+    algo_name_input: Option<&str>,
+    length_input: Option<usize>,
+    res: &mut ChecksumResult,
+    properly_formatted: &mut bool,
+) -> (String, Option<usize>) {
+    if is_algo_based_format {
```

Comment:
As most functionality is for the "algo based format", it might make sense to make this function "algo based format"-only and move the rest of the functionality somewhere else.